[HttpFoundation] Streamlining server event streaming

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

While [Mercure](https://mercure.rocks/) is the recommended option for most cases, implementing a custom Server-sent Event (SSE) backend relying on Symfony components is still a good alternative (considering you're building an app with a few users only)

Example using today’s code:
```php
public function __invoke(): StreamedResponse
{
    $response = new StreamedResponse(function () {
        foreach ($this->watchJobsInProgress() as $job) {
            echo "type: jobs\n";
            echo "data: ".$job->toJson()."\n\n";

            StreamedResponse::closeOutputBuffers(0, true);
            flush();

            if (connection_aborted()) {
                break;
            }

            sleep(1);
        }
    });
    $response->headers->set('Content-Type', 'text/event-stream');
    $response->headers->set('Cache-Control', 'no-cache');
    $response->headers->set('Connection', 'keep-alive');
    $response->headers->set('X-Accel-Buffering', 'no');

    return $response;
}
```

So, this PR simplifies that implementation by adding two new artifacts:

### ✨ New `ServerEvent` DTO class

Spec: https://html.spec.whatwg.org/multipage/server-sent-events.html

The `ServerEvent` class represents individual events streamed from the server to clients. It includes fields like `data`, `type`, `id`, `retry`, and `comment`, so it can cover a variety of use cases.

One key feature is the `getIterator()` method, which automatically concat all these fields into the proper SSE format. This saves us from having to manually format the events, making the process a lot smoother. Plus, it supports iterable `data`, meaning it can handle multi-line messages or more complex data structures all in one event.

### ✨ New `EventStreamResponse` class

* **Simplified Headers Setup** 
  
  ```php
  return new EventStreamResponse(/* ... */);
  ```
  This class automatically sets the required headers (`Content-Type`, `Cache-Control`, and `Connection`), ensuring that every response is properly configured for event streaming without additional code.

* **Generator-based Event Streaming with Auto Flush** 

  ```php
  return new EventStreamResponse(function (): \Generator {    
      yield new ServerEvent(time(), type: 'ping');
  }); 
  ```
  The callback function can now `yield` events directly, automatically handling serialization and output buffering as each new event occurs.

* **Event Streaming with Sending Control** 

  ```php
  return new EventStreamResponse(function (EventStreamResponse $response) {    
      $response->sendEvent(new ServerEvent('...'));
      // do something in between ...
      $response->sendEvent(new ServerEvent('...'));
  }); 
  ```
  The callback function receives a new argument—the response instance—allowing you to manually `sendEvent()`.
  
  This method is especially useful when `yield` is not possible due to scope limitations. In the example below, the callback handles the event inside an event stream response, listening for Redis messages and directly triggering the `sendEvent()` method:
  ```php
  return new EventStreamResponse(function (EventStreamResponse $response) {  
      $redis = new \Redis();  
      $redis->connect('127.0.0.1');  
      $redis->subscribe(['message'], function (/* ... */, string $message) use ($response) {  
          $response->sendEvent(new ServerEvent($message));  
      });  
  }); 
  ```

* **Retry Event Fallback**

  ```php
  return new EventStreamResponse(/* ... */, retry: 1000);
  ```
  `ServerEvent` allows setting individual `retry` intervals, which take precedence over the default retry configured in `EventStreamResponse`. If not specified at the event level, the global retry setting applies.

This is the final, optimized version after the proposed improvements:
```php
public function __invoke(): EventStreamResponse
{
    return new EventStreamResponse(function () {
        foreach ($this->watchJobsInProgress() as $job) {
            yield new ServerEvent($job->toJson(), type: 'jobs');

            sleep(1);
        }
    });
}
```
Cheers!

yceruto committed:
Streamlining server event streaming

yceruto commented:
@nicolas-grekas If we do it with the Symfony stack for sure, I agree with you, but we can still use the HttpFoundation component with ReactPHP, right?

nicolas-grekas commented:
Sure, but is that a real use case? How would such an app look like?

yceruto commented:
I don't have a real use case with ReactPHP, just with Symfony itself and an app within a private network. My infra is very small and I don't have the connections/memory problems in this case, just a few users.

My intention here is to simplify things for these straightforward cases; otherwise I'd recommend using Mercure of course.

yceruto commented:
I still think these improvements can significantly enhance the developer experience for small applications with a limited number of connections or where memory consumption is manageable.

Friendly ping @symfony/mergers is this something unconventional?

faizanakram99 commented:
Fwiw, I think it's useful too. We also have something similar in house mainly for live feedback, for interoperability with cli (symfony commands), it extends Output class (not ideal I know but fulfills the needs). 


```php
<?php

declare(strict_types=1);

namespace Zzz\Common;

use Psr\Log\LoggerInterface;
use Symfony\Component\Console\Output\Output;
use Symfony\Contracts\Service\Attribute\Required;

class EventSourceOutput extends Output
{
    #[Required]
    public LoggerInterface $logger;

    private bool $streamStarted = false;

    public function writeln(string|iterable $messages, int $options = self::OUTPUT_RAW): void
    {
        parent::writeln($messages, $options);
    }

    public function write(string|iterable $messages, bool $newline = false, int $options = self::OUTPUT_RAW): void
    {
        parent::write($messages, $newline, $options);
    }

    protected function doWrite(string $message, bool $newline): void
    {
        echo "{$this->stream($message)}\n\n";
    }

    private function stream(string $message): string
    {
        if (!$this->streamStarted) {
            $this->startStream();
        }

        if ((bool) \preg_match('|^<(\w+)>(.+)</(\w+)>$|', $message, $matches)) {
            return 'close' === $matches[1]
                ? "data: {$matches[2]}"
                : $this->prepareStream($matches[1], $matches[2]);
        }

        return $this->prepareStream('info', (string) \preg_replace('/\s+/', ' ', \strip_tags($message)));
    }

    private function prepareStream(string $type, string $text): string
    {
        if (!(bool) \preg_match('!!u', $text)) {
            // Not UTF-8
            $text = '...';
        }

        try {
            return \sprintf('data: %s', \json_encode(['type' => $type, 'text' => $text], \JSON_THROW_ON_ERROR));
        } catch (\JsonException $e) {
            $this->logger->critical($e->getMessage(), ['exception' => $e]);

            return 'data: END_SSE';
        }
    }

    private function startStream(): void
    {
        \ini_set('output_buffering', 'off');
        \ini_set('zlib.output_compression', '0');
        \ob_implicit_flush();

        while (\ob_get_level() > 0) {
            \ob_end_clean();
        }

        if (\function_exists('apache_setenv')) {
            apache_setenv('no-gzip', '1');
            apache_setenv('dont-vary', '1');
        }

        \header('Content-Type: text/event-stream');
        \header('Cache-Control: no-cache');
        \header('Connection: keep-alive');
        \header('X-Accel-Buffering: no');

        echo \str_pad('', 10000);
        echo "\n";

        $this->streamStarted = true;
    }

    public function hasStreamStarted(): bool
    {
        return $this->streamStarted;
    }

    public function endStream(): void
    {
        $this->writeln('<close>END_SSE</close>');
    }
}
```

Do note that `X-Accel-Buffering: no` header is needed for ngjnx, otherwise it buffers output

faizanakram99 commented on src/Symfony/Component/HttpFoundation/ServerEvent.php:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @author Yonel Ceruto <open@yceruto.dev>
+ */
+class ServerEvent
```
```suggestion
class ServerSideEvent
```

ServerSideEvent makes the intent clear that it represents an SSE payload.

yceruto replied:
thanks for the suggestion! I chose "ServerEvent" since SSE stands for "Server-Sent Event" and "ServerSideEvent" might be confused with an event that only occurs internally on the server. I updated the class description to make it clearer that it's an event generated on the server intended for streaming to the client, which then becomes a "Server-Sent Event"

welcoMattic commented:
I like this, as it allows developers of small/private projects with small traffic to leverage SSE without supporting a Mercure instance.

But, I also agree with Nicolas, this can quickly leads to high memory consumption and this should be well documented as a warning.

dunglas commented:
I'm in favor of this feature too. 

Mercure focus is event broadcasting, which cannot be (easily) achieved using this patch. But SSE is also useful for other use cases than broadcasting.

For instance, this feature could be useful for hinting the client about the progress of a synchronous job.

Regarding the memory/performance concerns, mentioning in the documentation that for broadcasting but also for high-traffic sites using Mercure would be more straightforward should be good enough.

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+ * Example usage:
+ *
+ *     return new EventStreamResponse(function () {
+ *         while (true) {
```
We should maybe not use an infinite loop in the docs, because PHP is not designed for that and this will crash (timeout, memory leaks...) except if php.ini is tweaked.

Maybe that we could just send some events with a sleep between them then close the connection.

yceruto replied:
agree, loop removed from the example

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+ * Example usage:
+ *
+ *     return new EventStreamResponse(function () {
+ *         while (true) {
```
We should maybe not use an infinite loop in the docs, because PHP is not designed for that and this will crash (timeout, memory leaks...) except if php.ini is tweaked.

Maybe that we could just send some events with a sleep between them then close the connection.

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Represents a streaming HTTP response for sending server events
+ * as part of the Server-Sent Events (SSE) streaming technique.
```
```suggestion
 * as part of the Server-Sent Events (SSE) streaming technique.
 *
 * To broadcast events to multiple users at once, for long-running connections and for high-traffic websites, prefer using the Mercure Symfony Component, which relies on software designed for these use cases: https://symfony.com/doc/current/mercure.html
```

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+class EventStreamResponse extends StreamedResponse
+{
+    /**
+     * @param int $retry The event reconnection time in milliseconds
```
```suggestion
     * @param int $retry The number of milliseconds the client should wait before reconnecting in case of network failure
```

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+        $headers += [
+            'Content-Type' => 'text/event-stream',
+            'Cache-Control' => 'no-cache',
+            'Connection' => 'keep-alive',
```
We should also set these legacy cache headers and the `X-Accel-Buffering` NGINX header by default to prevent issues with many reverse proxies (like NGINX Ingress for Kubernetes, which is very popular): https://github.com/dunglas/mercure/blob/main/subscribe.go#L218

yceruto replied:
Headers updated!

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @return $this
+     */
+    public function sendEvent(ServerEvent $event, bool $flush = true): static
```
What is the use case of not flushing?

yceruto replied:
I was thinking on sending multiple events together and reducing the number of separate writes (and thus TCP packets or system calls) but the benefit may be marginal, so I removed it.

dunglas commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+     */
+    public function sendEvent(ServerEvent $event, bool $flush = true): static
+    {
+        if ($this->retry > 0 && 0 === $event->getRetry()) {
```
You can set this attribute only when the value changes. It's not necessary to add it to all events.

yceruto replied:
done, thanks!

dunglas commented on src/Symfony/Component/HttpFoundation/ServerEvent.php:
```diff
@@ -0,0 +1,137 @@
+    public function __construct(
+        private string|iterable $data,
+        private ?string $type = null,
+        private int $retry = 0,
```
Maybe make it nullable too, as it's entirely optional by the spec.

dunglas commented on src/Symfony/Component/HttpFoundation/ServerEvent.php:
```diff
@@ -0,0 +1,137 @@
+        return $this;
+    }
+
+    public function __toString(): string
```
On my opinion, we should provide a `stream()` method that will stream the event in the HTTP response directly instead of storing the whole content of the text to be sent in memory.

This will be especially useful for large payloads.

yceruto replied:
hopefully it's better now. large payloads can use the `data` iterator property to set a traversable that will be streamed directly

jvasseur commented on src/Symfony/Component/HttpFoundation/EventStreamResponse.php:
```diff
@@ -0,0 +1,99 @@
+ *
+ *             if (connection_aborted()) {
+ *                 break;
+ *             }
```
We could make this check inside `EventStreamResponse` to reduce the code needed to implement an event stream.

yceruto replied:
good suggestion! It's part of internal loop now, thanks!

yceruto commented:
Happy to read your positive comments! I'll be addressing them soon, thanks!

yceruto commented:
I’ve addressed all the comments. hopefully, I understood them all correctly. 
I also updated the PR description with a more specific use case.

it's ready for review again 🤞

fabpot commented:
Thank you @yceruto.

welcoMattic commented:
Side note to mention that this feature will ease the integration of Symfony apps as [MCP servers](https://modelcontextprotocol.io/introduction), which as 2 transports modes: `stdio` and `SSE`. 🤖 cc @lyrixx
